### PR TITLE
feat:[CDS-73984]: Download Manifests Step Refactoring

### PR DESCRIPTION
**Pull request by piyush-bhuwalka**
## Describe your changes

Download Manifests step was implemented in such a way that it only categered to the needs of AWS SAM Swimlane. We wanted to generalize it and for that refactoring was done. 

https://harness.atlassian.net/wiki/spaces/CDNG/pages/21380465335/Approaches+with+GitClone

Tests done :
- Basic Sanity for SAM flow
- Testing for Serverless flow

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50357)

### CODE REVIEW

**Comment by piyush-bhuwalka** [comment]
trigger uts

**Comment by piyush-bhuwalka** [comment]
trigger messagecheck

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Review by RishabhGupta34 — DISMISSED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/DownloadManifestsStepHelper.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        ambiance, RefObjectUtils.getOutcomeRefObject(OutcomeExpressionConstants.SERVICE));
+
+    if (ServiceSpecType.AWS_SAM.equals(serviceOutcome.getType())) {
+      checkSamFeatureFlagEnabled(ambiance);
+      return awsSamDownloadManifestsStepHelper.handleAsyncResponse(ambiance, responseDataMap);
+    } else if (ServiceSpecType.SERVERLESS_AWS_LAMBDA.equals(serviceOutcome.getType())) {
+      checkServerlessV2FeatureFlagEnabled(ambiance);
+      return serverlessDownloadManifestsV2StepHelper.handleAsyncResponse(ambiance, responseDataMap);
+    } else {
+      return null;
```
Comment (anilchowdhury-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/DownloadManifestsStepHelper.java:75]: throw error .. stating Download manifest is not supported for this specific service type

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    ServerlessAwsLambdaManifestOutcome serverlessAwsLambdaDirectoryManifestOutcome =
+        (ServerlessAwsLambdaManifestOutcome) getServerlessAwsLambdaDirectoryManifestOutcome(manifestsOutcome.values());
+
+    List<String> callbackIds = new ArrayList<>();
+    List<String> logKeys = new ArrayList<>();
+
+    GitCloneStepInfo gitCloneStepInfo = downloadManifestsCommonHelper.getGitCloneStepInfoFromManifestOutcome(
+        serverlessAwsLambdaDirectoryManifestOutcome);
+
+    StepElementParameters stepElementParameters = downloadManifestsCommonHelper.getGitStepElementParameters(
```
Comment (anilchowdhury-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java:89]: refactor code to look something like this

downloadManifest()
downloadValues()
renderValues()

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+          StepStatusTaskResponseData stepStatusTaskResponseData = (StepStatusTaskResponseData) responseData;
+          if (stepStatusTaskResponseData != null
+              && stepStatusTaskResponseData.getStepStatus().getStepExecutionStatus() == StepExecutionStatus.SUCCESS) {
+            StepOutput stepOutput = stepStatusTaskResponseData.getStepStatus().getOutput();
+
+            if (stepOutput instanceof StepMapOutput) {
+              String valuesYamlPath = getValuesPathFromValuesManifestOutcome(valuesManifestOutcome);
+
+              StepMapOutput stepMapOutput = (StepMapOutput) stepOutput;
+              if (stepMapOutput.getMap() != null && stepMapOutput.getMap().size() > 0) {
```
Comment (anilchowdhury-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java:155]: lets use isNotEmpty()

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    pluginCreationResponseWrapperList.add(pluginCreationResponseWrapper);
+
+    // Values Yaml
+
+    ValuesManifestOutcome valuesManifestOutcome =
+        (ValuesManifestOutcome) getServerlessAwsLambdaValuesManifestOutcome(manifestsOutcome.values());
+
+    if (valuesManifestOutcome != null) {
+      GitCloneStepInfo valuesGitCloneStepInfo =
+          downloadManifestsCommonHelper.getGitCloneStepInfoFromManifestOutcomeWithOutputFilePathContents(
```
Comment (anilchowdhury-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java:231]: lets move this to method
renderValueManifestAndSaveToSweepingOutCome()

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    ManifestsOutcome manifestsOutcome =
+        (ManifestsOutcome) outcomeService
+            .resolveOptional(ambiance, RefObjectUtils.getOutcomeRefObject(OutcomeExpressionConstants.MANIFESTS))
+            .getOutcome();
+
+    ServerlessAwsLambdaManifestOutcome serverlessAwsLambdaManifestOutcome =
+        (ServerlessAwsLambdaManifestOutcome) getServerlessAwsLambdaDirectoryManifestOutcome(manifestsOutcome.values());
+
+    GitCloneStepInfo gitCloneStepInfo =
+        downloadManifestsCommonHelper.getGitCloneStepInfoFromManifestOutcome(serverlessAwsLambdaManifestOutcome);
```
Comment (anilchowdhury-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java:209]: lets move to a separate method

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/AwsSamDownloadManifestsStepHelper.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
+@Slf4j
+public class AwsSamDownloadManifestsStepHelper {
```
Comment (anilchowdhury-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/AwsSamDownloadManifestsStepHelper.java:61]: refactor into smaller method as commented for ServerlessManifestHelper

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java (hunk 3 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java:155]: Done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java (hunk 2 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java]: Refactored but with bit different method names

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
+@Slf4j
+public class ServerlessDownloadManifestsV2StepHelper {
+  @Inject private OutcomeService outcomeService;
```
[the remark this answers is not in the document]
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java]: Done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/DownloadManifestsStepHelper.java (hunk 1 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/DownloadManifestsStepHelper.java]: Done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java (hunk 4 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java]: Done

**Review by anilchowdhury-harness — APPROVED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java (hunk 5 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/ServerlessDownloadManifestsV2StepHelper.java]: Done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/AwsSamDownloadManifestsStepHelper.java (hunk 6 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/downloadmanifests/AwsSamDownloadManifestsStepHelper.java]: Done

**Review by anilchowdhury-harness — APPROVED** [review]

**Comment by piyush-bhuwalka** [comment]
trigger uts

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 21 files at r1, 1 of 1 files at r2, 10 of 20 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50357)* status: 24 file(s) to review. 7. unresolved discussions. 0 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]

**Review by anilchowdhury-harness — APPROVED** [review]

**Comment by piyush-bhuwalka** [comment]
trigger review

**Pull request merged**